File: factory/production_lines/android/templates/analytics/AnalyticsEvents.kt
package {{PACKAGE_NAME}}.analytics

import android.os.Bundle
import androidx.core.os.bundleOf

/**
 * DriveAI Standard Analytics Events — Sealed Class Hierarchy
 *
 * Event names are IDENTICAL to iOS for cross-platform consistency.
 * All events use the "dai_" prefix.
 *
 * Usage:
 *   val event = AnalyticsEvents.Session.AppOpen
 *   AnalyticsManager.logEvent(event.eventName, event.parameters)
 */
sealed class AnalyticsEvents(
    val eventName: String,
    val parameters: Bundle? = null
) {

    // ── Session Events ──────────────────────────────────────────────

    sealed class Session(eventName: String, parameters: Bundle? = null) :
        AnalyticsEvents(eventName, parameters) {

        /** App opened / foregrounded */
        object AppOpen : Session("dai_app_open")

        /** App moved to background */
        object AppBackground : Session("dai_app_background")
    }

    // ── Onboarding Funnel ───────────────────────────────────────────

    sealed class Onboarding(eventName: String, parameters: Bundle? = null) :
        AnalyticsEvents(eventName, parameters) {

        /** User starts onboarding flow */
        object Start : Onboarding("dai_onboarding_start")

        /** User completes a specific onboarding step */
        data class Step(val step: Int, val stepName: String) : Onboarding(
            eventName = "dai_onboarding_step",
            parameters = bundleOf(
                "step_number" to step,
                "step_name" to stepName
            )
        )

        /** User completes entire onboarding */
        object Complete : Onboarding("dai_onboarding_complete")

        /** User skips onboarding */
        data class Skip(val atStep: Int) : Onboarding(
            eventName = "dai_onboarding_skip",
            parameters = bundleOf("skipped_at_step" to atStep)
        )
    }

    // ── Feature Usage ───────────────────────────────────────────────

    sealed class Feature(eventName: String, parameters: Bundle? = null) :
        AnalyticsEvents(eventName, parameters) {

        /** User actively uses a feature */
        data class Used(val featureName: String) : Feature(
            eventName = "dai_feature_used",
            parameters = bundleOf("feature_name" to featureName)
        )

        /** User discovers a feature (first interaction) */
        data class Discovered(val featureName: String) : Feature(
            eventName = "dai_feature_discovered",
            parameters = bundleOf("feature_name" to featureName)
        )
    }

    // ── Engagement ──────────────────────────────────────────────────

    sealed class Engagement(eventName: String, parameters: Bundle? = null) :
        AnalyticsEvents(eventName, parameters) {

        /** Active session milestone (e.g. 1min, 5min, 10min) */
        data class SessionActive(val durationSeconds: Long) : Engagement(
            eventName = "dai_session_active",
            parameters = bundleOf("duration_seconds" to durationSeconds)
        )

        /** User views specific content */
        data class ContentViewed(val contentType: String, val contentId: String) : Engagement(
            eventName = "dai_content_viewed",
            parameters = bundleOf(
                "content_type" to contentType,
                "content_id" to contentId
            )
        )
    }

    // ── Monetization ────────────────────────────────────────────────

    sealed class Monetization(eventName: String, parameters: Bundle? = null) :
        AnalyticsEvents(eventName, parameters) {

        /** User initiates a purchase */
        data class PurchaseStart(val productId: String, val value: Double? = null) : Monetization(
            eventName = "dai_purchase_start",
            parameters = bundleOf(
                "product_id" to productId
            ).apply {
                value?.let { putDouble("value", it) }
            }
        )

        /** Purchase completed successfully */
        data class PurchaseComplete(
            val productId: String,
            val value: Double,
            val currency: String
        ) : Monetization(
            eventName = "dai_purchase_complete",
            parameters = bundleOf(
                "product_id" to productId,
                "value" to value,
                "currency" to currency
            )
        )

        /** Subscription started */
        data class SubscriptionStart(
            val planId: String,
            val value: Double,
            val currency: String
        ) : Monetization(
            eventName = "dai_subscription_start",
            parameters = bundleOf(
                "plan_id" to planId,
                "value" to value,
                "currency" to currency
            )
        )

        /** Ad impression tracked */
        data class AdImpression(
            val adUnit: String,
            val adFormat: String? = null
        ) : Monetization(
            eventName = "dai_ad_impression",
            parameters = bundleOf(
                "ad_unit" to adUnit
            ).apply {
                adFormat?.let { putString("ad_format", it) }
            }
        )
    }

    // ── Errors ──────────────────────────────────────────────────────

    sealed class Error(eventName: String, parameters: Bundle? = null) :
        AnalyticsEvents(eventName, parameters) {

        /** Non-fatal error occurred */
        data class Occurred(
            val errorCode: String,
            val errorMessage: String,
            val screen: String? = null
        ) : Error(
            eventName = "dai_error_occurred",
            parameters = bundleOf(
                "error_code" to errorCode,
                "error_message" to errorMessage
            ).apply {
                screen?.let { putString("screen", it) }
            }
        )
    }
}

File: factory/production_lines/web/templates/analytics/analyticsEvents.ts
// analyticsEvents.ts
// DAI-Core Analytics Event Definitions — Web
//
// Type-safe event factory functions for all standard DAI events.
// Event names are cross-platform identical with iOS/Android.
// All events use the "dai_" prefix.

// ── Types ────────────────────────────────────────────────────────

/** A structured analytics event ready to be logged. */
export interface AnalyticsEvent {
  name: string;
  parameters?: Record<string, any>;
}

// ── Lifecycle Events ─────────────────────────────────────────────

/** App opened / page loaded. */
export function daiAppOpen(): AnalyticsEvent {
  return { name: "dai_app_open" };
}

/** App moved to background (visibilitychange / pagehide). */
export function daiAppBackground(): AnalyticsEvent {
  return { name: "dai_app_background" };
}

// ── Onboarding Events ────────────────────────────────────────────

/** User started the onboarding flow. */
export function daiOnboardingStart(): AnalyticsEvent {
  return { name: "dai_onboarding_start" };
}

/** User reached a specific onboarding step. */
export function daiOnboardingStep(
  step: number,
  stepName: string
): AnalyticsEvent {
  return {
    name: "dai_onboarding_step",
    parameters: {
      step_number: step,
      step_name: stepName,
    },
  };
}

/** User completed the entire onboarding flow. */
export function daiOnboardingComplete(): AnalyticsEvent {
  return { name: "dai_onboarding_complete" };
}

/** User skipped the onboarding flow. */
export function daiOnboardingSkip(atStep?: number): AnalyticsEvent {
  return {
    name: "dai_onboarding_skip",
    parameters: atStep !== undefined ? { skipped_at_step: atStep } : undefined,
  };
}

// ── Feature Events ───────────────────────────────────────────────

/** User actively used a feature. */
export function daiFeatureUsed(featureName: string): AnalyticsEvent {
  return {
    name: "dai_feature_used",
    parameters: { feature_name: featureName },
  };
}

/** User discovered a feature (first interaction / tooltip shown). */
export function daiFeatureDiscovered(featureName: string): AnalyticsEvent {
  return {
    name: "dai_feature_discovered",
    parameters: { feature_name: featureName },
  };
}

// ── Session Events ───────────────────────────────────────────────

/** Session heartbeat — user is actively using the app. */
export function daiSessionActive(
  durationSeconds: number
): AnalyticsEvent {
  return {
    name: "dai_session_active",
    parameters: { duration_seconds: durationSeconds },
  };
}

// ── Content Events ───────────────────────────────────────────────

/** User viewed a piece of content. */
export function daiContentViewed(
  contentType: string,
  contentId: string
): AnalyticsEvent {
  return {
    name: "dai_content_viewed",
    parameters: {
      content_type: contentType,
      content_id: contentId,
    },
  };
}

// ── Monetization Events ──────────────────────────────────────────

/** User started a purchase flow. */
export function daiPurchaseStart(
  itemId: string,
  value?: number,
  currency?: string
): AnalyticsEvent {
  const parameters: Record<string, any> = { item_id: itemId };
  if (value !== undefined) parameters.value = value;
  if (currency !== undefined) parameters.currency = currency;
  return {
    name: "dai_purchase_start",
    parameters,
  };
}

/** User completed a purchase. */
export function daiPurchaseComplete(
  itemId: string,
  value: number,
  currency: string,
  transactionId?: string
): AnalyticsEvent {
  const parameters: Record<string, any> = {
    item_id: itemId,
    value,
    currency,
  };
  if (transactionId !== undefined) parameters.transaction_id = transactionId;
  return {
    name: "dai_purchase_complete",
    parameters,
  };
}

/** User started a subscription. */
export function daiSubscriptionStart(
  planId: string,
  value?: number,
  currency?: string
): AnalyticsEvent {
  const parameters: Record<string, any> = { plan_id: planId };
  if (value !== undefined) parameters.value = value;
  if (currency !== undefined) parameters.currency = currency;
  return {
    name: "dai_subscription_start",
    parameters,
  };
}

/** Ad impression recorded. */
export function daiAdImpression(
  adUnit: string,
  adFormat?: string
): AnalyticsEvent {
  const parameters: Record<string, any> = { ad_unit: adUnit };
  if (adFormat !== undefined) parameters.ad_format = adFormat;
  return {
    name: "dai_ad_impression",
    parameters,
  };
}

// ── Error Events ─────────────────────────────────────────────────

/** An error occurred in the app. */
export function daiErrorOccurred(
  errorCode: string,
  errorMessage: string,
  isFatal?: boolean
): AnalyticsEvent {
  return {
    name: "dai_error_occurred",
    parameters: {
      error_code: errorCode,
      error_message: errorMessage,
      is_fatal: isFatal ?? false,
    },
  };
}
